### PR TITLE
executor: support batch insert for insert values statement

### PR DESCRIPTION
**Pull request by jackysp**
### What problem does this PR solve? 
For large `insert into values` statement (insert many values in one statement), memory usage is high for the session holds all rows in one statement.

### What is changed and how it works?
Support batch insert support for `insert into values` statement.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - Increased code complexity

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8420)

### CODE REVIEW

**Comment by tiancaiamao** [comment]
> For large insert statement (insert many values in one statement), memory usage is high for the session holds all rows in one statement.

Changes in this PR do not reduce memory usage. @jackysp 

> Increased code complexity

That's true.

File: executor/builder.go [hunk 1]
```diff
@@ -552,15 +552,19 @@ func (b *executorBuilder) buildInsert(v *plannercore.Insert) Executor {
...
+		baseExecutor:  baseExec,
+		Table:         v.Table,
+		Columns:       v.Columns,
+		Lists:         v.Lists,
+		SetList:       v.SetList,
+		GenColumns:    v.GenCols.Columns,
+		GenExprs:      v.GenCols.Exprs,
+		hasRefCols:    v.NeedFillDefaultValue,
+		SelectExec:    selectExec,
+		rowCountLimit: uint64(b.ctx.GetSessionVars().DMLBatchSize / len(v.Table.Cols())),
```
Comment (zz-jason) [inline executor/builder.go]: why divide `len(v.Table.Cols())`?

**Comment by zz-jason** [comment]
@lysu @tiancaiamao PTAL

File: executor/builder.go (hunk 1 above)
Comment (jackysp) [reply executor/builder.go]: It could balance the `rowCountLimit` with the table width.

**Comment by jackysp** [comment]
> > For large insert statement (insert many values in one statement), memory usage is high for the session holds all rows in one statement.
> 
> Changes in this PR do not reduce memory usage. @jackysp
> 
> > Increased code complexity
> 
> That's true.

"Changes in this PR do not reduce memory usage."
Are you sure about it? @tiancaiamao

**Comment by zz-jason** [comment]
Is there any benchmark result?

**Comment by jackysp** [comment]
Yes, we could keep on optimizing it. This pr also balance the batch size for the wide and the narrow table, and make batch insert work on `insert into values` statement. I'll update the description.

**Comment by jackysp** [comment]
> If this PR is aimed to reduce the memory consumption, we can bench the memory usage.
> If this PR is aimed to reduce the execution time, we can bench the execution time.

These are been done in the customer's envriment. I can not post them here for the security reasons.

File: executor/insert_common.go [hunk 2]
```diff
@@ -328,6 +341,19 @@ func (e *InsertValues) insertRowsFromSelect(ctx context.Context, exec func(rows
 	return nil
 }
 
+func (e *InsertValues) doBatchInsert() error {
+	sessVars := e.ctx.GetSessionVars()
+	e.ctx.StmtCommit()
```
Comment (tiancaiamao) [inline executor/insert_common.go]: StmtCommit should be done only when the statement finish.
Batch insert during a statement is executing is not safe. @jackysp

Comment (jackysp) [reply executor/insert_common.go]: It must be done. Due to it is going to start a new transaction.

**Comment by tiancaiamao** [comment]
> > If this PR is aimed to reduce the memory consumption, we can bench the memory usage.
> > If this PR is aimed to reduce the execution time, we can bench the execution time.
> 
> These are been done in the customer's envriment. I can not post them here for the security reasons.

I still in doubt about it.

File: executor/insert_common.go (hunk 2 above)
Comment (tiancaiamao) [reply executor/insert_common.go]: ok, this function runs when `batchInsert` is set.

**Comment by tiancaiamao** [comment]
I come up with another problem in this change:

`insert into v values (1),(2),(3),(4),(5).....(1000), (1001)...(10000)`

If batchInsert commit  (1)....(1000), and meet error later, would the TiDB retry this statement ? 
Is it possible that half of the statement (1) ... (1000) is succeed ?

**Comment by jackysp** [comment]
I think it could not retry. And it is added by you @tiancaiamao .  #5745

File: executor/builder.go (hunk 1 above)
Comment (winkyao) [reply executor/builder.go]: What if DMLBatchSize less than `len(v.Table.Cols())`?

Comment (jackysp) [reply executor/builder.go]: It will be set to `MinRowCountLimit`

Comment (winkyao) [reply executor/builder.go]: Fine, rowCounLimit is at least MinRowCountLimit

**Comment by winkyao** [comment]
What @tiancaiamao said, seems like reasonable, are sure this PR can reduce the memory?

**Comment by jackysp** [comment]
It's not worked as expected, I've updated the title and the description.

File: executor/insert_common.go [hunk 3]
```diff
@@ -306,18 +322,15 @@ func (e *InsertValues) insertRowsFromSelect(ctx context.Context, exec func(rows
 				return errors.Trace(err)
 			}
 			rows = append(rows, row)
-			if batchInsert && e.rowCount%uint64(batchSize) == 0 {
-				if err := exec(rows); err != nil {
+			if e.rowCount%e.rowCountLimit == 0 {
```
Comment (tiancaiamao) [inline executor/insert_common.go]: I think you should update sessionVar comment for TiDBDMLBatchSize, if you change here.

Comment (tiancaiamao) [reply executor/insert_common.go]: Or use batchSize instead of rowCountLimit

Comment (jackysp) [reply executor/insert_common.go]: The comment of `TiDBDMLBatchSize` doesn't show it is a row size. I think this change won't mislead the comment reader.

**Comment by jackysp** [comment]
PTAL @tiancaiamao

File: executor/insert_common.go [hunk 4]
```diff
@@ -188,6 +191,17 @@ func (e *InsertValues) insertRows(exec func(rows [][]types.Datum) error) (err er
 			return errors.Trace(err)
 		}
 		rows = append(rows, row)
+		if e.rowCount%uint64(batchSize) == 0 {
+			if err = exec(rows); err != nil {
+				return err
+			}
+			rows = rows[:0]
+			if batchInsert {
```
Comment (tiancaiamao) [inline executor/insert_common.go]: Together with line 183, the changes make `begin ... commit` can't commit midway

**Comment by jackysp** [comment]
PTAL @tiancaiamao

**Comment by tiancaiamao** [comment]
LGTM @winkyao @zz-jason

File: executor/insert_common.go (hunk 4 above)
Comment (tiancaiamao) [reply executor/insert_common.go]: Yep, the code is here since the first day it's written

**Comment by zz-jason** [comment]
Em, how much memory can be reduced in insert after this PR?

**Pull request merged**